File: 30. Substring with Concatenation of All Words.cpp
// https://leetcode.com/problems/substring-with-concatenation-of-all-words/

// You are given a string s and an array of strings words of the same length.
// Return all starting indices of substring(s) in s that is a concatenation of each
// word in words exactly once, in any order, and without any intervening
// characters.

// You can return the answer in any order.

////////////////////////////////////////////////////////////////////////////////

// sliding window
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        unordered_map<string, int> target;
        for (string& word : words) ++target[word];
        
        vector<int> ans;
        int n = s.size(), k = words[0].size();
        // check s.substr(offset, m * k) == target
        for (int offset = 0; offset < k; ++offset) {
            // sliding window
            unordered_map<string, int> window;
            int left = offset, right = offset, match = 0;
            while (right + k - 1 < n) {
                string newWord = s.substr(right, k);
                if (target.find(newWord) != target.end()) { // newWord is good
                    ++window[newWord];
                    if (window[newWord] == target[newWord]) ++match;
                } else { // skip bad word, reset window
                    window.clear(); match = 0; right += k; left = right;
                    continue;
                }
                // shrink window
                while (window[newWord] > target[newWord]) {
                    string oldWord = s.substr(left, k);
                    if (window[oldWord] == target[oldWord]) --match;
                    --window[oldWord];
                    left += k;
                }
                if (match == target.size()) ans.push_back(left); // update ans
                right += k; // expand window
            }
        }
        return ans;
    }
};

File: Python/30. Substring with Concatenation of All Words.py
# https://leetcode.com/problems/substring-with-concatenation-of-all-words/

# You are given a string s and an array of strings words of the same length.
# Return all starting indices of substring(s) in s that is a concatenation of
# each word in words exactly once, in any order, and without any intervening
# characters.

# You can return the answer in any order.

###############################################################################

# sliding window, use dict for comparison
# check s[0 : m * k] == words
# move sliding window -> check s[k : m * k + k] == words
# loop over offset in range(k) -> check s[offset : offset + m * k] == words

from collections import defaultdict
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        n, m, k = len(s), len(words), len(words[0])
        
        # init words dict
        words_dict = defaultdict(int)
        for word in words:
            words_dict[word] += 1

        ans = []
        running_dict = defaultdict(int)
        for offset in range(k):
            running_dict.clear()
            
            # init sliding window
            left, right = offset, offset + m * k
            if right > n: return ans
            
            for i in range(left, right, k):
                word = s[i : i + k]
                running_dict[word] += 1
            
            if running_dict == words_dict:
                ans.append(left)

            # move sliding window
            for i in range(right, n, k):
                next_word = s[i : i + k]
                prev_word = s[i - m * k : i - m * k + k]
                
                running_dict[next_word] += 1
                running_dict[prev_word] -= 1
                if running_dict[prev_word] == 0:
                    running_dict.pop(prev_word)
                
                if running_dict == words_dict:
                    ans.append(i - m * k + k)
        
        return ans